0006-ffmul-revised.md

```diff
@@ -0,0 +1,1932 @@
+t + \log_2(n) &> \log_2(2^{2m}) = 2m
+\end{aligned}
+$$
+
+So we get
+
+$$
+m_{max} = \frac{t + \log_2(n)}{2}.
+$$
+
+With $t=258, n=255$ we have 
+
+$$
+\begin{aligned}
+m_{max} &= \frac{258 + 255}{2} = 256.5,
+\end{aligned}
+$$
+
+which is not enough space to handle anything larger than 256 bit moduli.  Instead, we will use $t=264$, giving $m_{max} = 259$ bits.
+
+The above formula is useful for checking the maximum number of bits supported of the foreign field modulus, but it is not useful for computing the maximum foreign field modulus itself (because $2^{m_{max}}$ is too coarse). For these checks, we can compute our maximum foreign field modulus more precisely with
+
+$$
+max_{mod} = \lfloor \sqrt{2^t \cdot n} \rfloor.
+$$
+
+The max prime foreign field modulus satisfying the above inequality for both Pallas and Vesta is `926336713898529563388567880069503262826888842373627227613104999999999999999607`.
+
+### Limb configuration
+
+Choosing the right limb size and the right number of limbs is a careful balance between the number of constraints (i.e. the polynomial degree) and the witness length (i.e. the number of rows).  Because one limiting factor that we have in Kimchi is the 12-bit maximum for range check lookups, we could be tempted to introduce 12-bit limbs.  However, this would mean having many more limbs, which would consume more witness elements and require significantly more rows.  It would also increase the polynomial degree by increasing the number of constraints required for the *intermediate products* (more on this later).
+
+We need to find a balance between the number of limbs and the size of the limbs.  The limb configuration is dependent on the value of $t$ and our maximum foreign modulus (as described in the previous section).  The larger the maximum foreign modulus, the more witness rows we will require to constrain the computation.  In particular, each limb needs to be constrained by the range check gate and, thus, must be in a copyable (i.e. permuteable) witness cell.  We have at most 7 copyable cells per row and gates can operate on at most 2 rows, meaning that we have an upperbound of at most 14 limbs per gate (or 7 limbs per row).
+
+As stated above, we want the foreign field modulus to fit in as few rows as possible and we need to constrain operands $a, b$, the quotient $q$ and remainder $r$. Each of these will require cells for each limb.  Thus, the number of cells required for these is
+
+$$
+cells = 4 \cdot limbs
+$$
+
+It is highly advantageous for performance to constrain foreign field multiplication with the minimal number of gates. This not only helps limit the number of rows, but also to keep the gate selector polynomial small. Because of all of this, we aim to constrain foreign field multiplication with a single gate (spanning at most $2$ rows). As mentioned above, we have a maximum of 14 permuteable cells per gate, so we can compute the maximum number of limbs that fit within a single gate like this.
```

Comment:
> keep the gate selector polynomial small
What does this mean? Sentence should probably be deleted

Reply:
Suggestion: replace the whole final sentence with
> We can compute the maximum number of limbs that fit in the 14 permutable columns as follows:

Reply:
> keep the gate selector polynomial small
I guess this refers to having  a gate that has small degree, but yeah not super clear. Delete.